[release-1.29][go1.23] hack: bump go-tools and exclude SA1006

MadhavJivrajani opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Backporting https://github.com/kubernetes/kubernetes/pull/126709 and https://github.com/kubernetes/kubernetes/pull/125926


#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```
/sig architecture testing
/assign @dims @liggitt

dims commented:
/approve
/lgtm

logicalhan commented:
/triage accepted

saschagrunert commented:
/retest

MadhavJivrajani commented:
/retest

saschagrunert commented:
/retest

mepi262 commented:
@MadhavJivrajani
It seems that some tests are failed.
Would you check and resolve it?

mepi262 commented:
/retest

dims commented:
/approve
/lgtm
/retest

MadhavJivrajani commented:
Taking a look

MadhavJivrajani committed:
hack/tools: bump go-tools for go1.23.x

Signed-off-by: Madhav Jivrajani <madhav.jiv@gmail.com>

MadhavJivrajani committed:
hack: disable SA1006 check

Signed-off-by: Madhav Jivrajani <madhav.jiv@gmail.com>

MadhavJivrajani committed:
.*: fix lint errors
Fix SA4032 and SA3000 which are flagged after bumping to go1.23
and v0.5.1 of go-tools.

Signed-off-by: Madhav Jivrajani <madhav.jiv@gmail.com>

MadhavJivrajani commented:
The failing job is a little bit of a chicken and egg problem:

- Bumping go-tools to v0.5.1 forces the go.mod in hack/tools to be atleast go1.22.1
- However, our CI runs tests with go1.21.4 (which is what K8s 1.29 was first released with in order to ensure things are still compatible)
- Since go1.21, the value of the `go` directive in go.mod enforces a minimum go version. 
  - Since go-tools is now at 1.22.x, the minimum go version required is that and testing with go1.21.x in our CI won't really work. 

Maybe it makes sense to edit our CI job to test out with go1.22.x instead since we already have K8s 1.29 patch released that were released on go1.22.x?

liggitt commented:
> * However, our CI runs tests with go1.21.4 (which is what K8s 1.29 was first released with in order to ensure things are still compatible)

we only run unit tests and integration tests that way... what happens if you bump `hack/tools/go.mod` to go 1.22?

MadhavJivrajani commented:
> we only run unit tests and integration tests that way... what happens if you just bump hack/tools/go.mod to go 1.22?

@liggitt here's logs for when I did this: https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/126713/pull-kubernetes-unit-go-compatibility/1873842915755692032

What I think is happening is, while we do run only unit and integration with go1.21 for the compat jobs, the make rule seems to be calling into installing tools:
https://github.com/kubernetes/kubernetes/blob/b7ef173c59065f9a5f68eb514ef0483c6f3887ae/hack/make-rules/test.sh#L182-L198
which then causes the issue of 
```
go: go.mod requires go >= 1.22.1 (running go 1.21.4; GOTOOLCHAIN=go1.21.4)
```

liggitt commented:
looks like this is blocked on https://github.com/kubernetes/kubernetes/pull/129566, would be good to get that wrapped up and verify it lets us bump just hack/tools/go.mod to go 1.22.1

liggitt commented:
opened https://github.com/kubernetes/kubernetes/pull/130171 attempting to combine this with a hack/tools go.mod bump and https://github.com/kubernetes/kubernetes/pull/129566 to get CI green

liggitt commented:
https://github.com/kubernetes/kubernetes/pull/130171 is green, closing in favor of that PR

/close
